### ISSUE TITLE
Mismatched lifetime on trait impl shows a misleading error

### ISSUE THREAD

**Issue by Earlz** [rust-lang/rust#80701]
The rust compiler gives a very poor error message when there is a lifetime expectation mismatch on trait usage.

I used this code: (playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=581fa159f8ff99f2df447308fccd4bbc )

```rust
use std::cell::*;

#[derive(Default)]
struct Test {
    pub foo: u32,
}

trait FooSetter {
    fn set_foo(&mut self, value: u32);
}

impl FooSetter for Test {
    fn set_foo(&mut self, value: u32) {
        self.foo = value;
    }
}

trait BaseSetter{
    fn set(&mut self, value: u32);
}
impl BaseSetter for dyn FooSetter{
    fn set(&mut self, value: u32){
        self.set_foo(value);
    }
}

struct TestHolder<'a> {
    pub holder: Option<RefCell<&'a mut dyn FooSetter>>,
}

impl <'a>TestHolder<'a>{
    pub fn test_foo(&self){
       self.holder.as_ref().unwrap().borrow_mut().set(20); 
    }
}



fn main() {
    let mut test = Test::default();
    test.foo = 10;
    {
        let holder = TestHolder { holder: Some(RefCell::from(&mut test))};
        
        holder.test_foo();
    }
    test.foo = 30;
}

```

I expected to see a useful error message, such as "expected trait `BaseSetter<'a>`, found trait: `BaseSetter<'static>`"

Instead, I was greeted with this almost useless error message: "expected trait `BaseSetter`, found trait `BaseSetter`"

The same error occurs on stable and nightly compilers

**Comment by Earlz** [comment]
For reference, this error message is resolved by changing 

```rust
impl BaseSetter for dyn FooSetter
```

To this explicit lifetime scoped version:

```rust
impl <'a>BaseSetter for (dyn FooSetter + 'a)
```

**Comment by Dylan-DPC** [comment]
Current error:

```
error[[E0521]](https://doc.rust-lang.org/stable/error_codes/E0521.html): borrowed data escapes outside of associated function
  --> src/main.rs:33:8
   |
31 | impl <'a>TestHolder<'a>{
   |       -- lifetime `'a` defined here
32 |     pub fn test_foo(&self){
   |                     ----- `self` is a reference that is only valid in the associated function body
33 |        self.holder.as_ref().unwrap().borrow_mut().set(20); 
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |        |
   |        `self` escapes the associated function body here
   |        argument requires that `'a` must outlive `'static`
   |
   = note: requirement occurs because of a mutable reference to `dyn FooSetter`
   = note: mutable references are invariant over their type parameter
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
   ```
